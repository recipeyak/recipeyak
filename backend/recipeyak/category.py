from __future__ import annotations

from collections import defaultdict
from functools import cache
from typing import Any

from recipeyak.inflect import singularize

_DEPARTMENT_MAPPING = {
    "alcohol": {
        "wine",
        "beer",
        "rum",
        "vodka",
        "vermouth",
        "amaretto",
        "mirin",
        "bourbon",
        "brandy",
        "scotch",
        "cognac",
        "sherry",
        "guinness",
        "stout",
        "madeira",
        "marsala",
    },
    "produce": {
        "basil",
        "coconut",
        "blueberries",
        "blackberries",
        "raspberries",
        "chile",
        "chili",
        "lemongrass",
        "eggplant",
        "cubanelle peppers",
        "plum",
        "apricots",
        "sage",
        "tarragon",
        "yamaimo",
        "romaine",
        "habanero chile",
        "snow pea",
        "spring mix",
        "dates",
        "prunes",
        "corn",
        "tomatillos",
        "salad",
        "pico de gallo",
        "peas",
        "bean sprouts",
        "passion fruit",
        "mushroom",
        "herbs",
        "butternut squash",
        "pasilla chilies",
        "mulato chiles",
        "haricot verts",
        "haricots verts",
        "greens",
        "pomegranate",
        "edamame",
        "rhubarb",
        "zucchini",
        "watermelon",
        "brussel sprouts",
        "brussels sprouts",
        "brusel sprouts",
        "dill",
        "beets",
        "broccoli",
        "berries",
        "leeks",
        "guacamole",
        "ancho chiles",
        "chipotle chiles",
        "pequin chile",
        "cascabel chile",
        "arbor chile",
        "chile negro",
        "chile ancho",
        "cauliflower",
        "fennel",
        "cranberries",
        "scallion",
        "bok choy",
        "turnip",
        "red chili",
        "tofu",
        "potato",
        "pear",
        "strawberries",
        "strawberris",
        "strawberry",
        "grapefruit",
        "serrano chile",
        "thai chile",
        "green chillies",
        "green chiles",
        "bird eye chillies",
        "carrot",
        "chive",
        "avocado",
        "watercress",
        "cabbage",
        "scallions",
        "lime",
        "thyme",
        "radish",
        "lemon",
        "tomato",
        "cherry",
        "grape",
        "lettuce",
        "shallot",
        "rosemary",
        "apple",
        "pineapple",
        "banana",
        "kale",
        "cilantro",
        "asparagus",
        "spinach",
        "orange",
        "arugula",
        "jalapeno",
        "jalapeño",
        "bell pepper",
        "poblano pepper",
        "fresno pepper",
        "onion",
        "parsley",
        "parsely",
        "mint",
        "garlic",
        "cloves garlic",
        "ginger",
        "clementine",
        "celery",
        "oregano",
        "cucumber",
    },
    "meat": {
        "chicken",
        "shrimp",
        "mussels",
        "beef",
        "pepperoni",
        "pork",
        "bacon",
        "sausage",
        "salmon",
        "kielbasa",
        "merguez",
        "chuck-eye roast",
        "pancetta",
        "oxtails",
        "ham",
        "chuck roast",
        "round roast",
        "bratwurst",
        "steak",
        "prawns",
        "turkey",
        "veal",
        "sirloin",
        "chorizo",
        "short ribs",
        "shortribs",
        "lamb",
        "fish",
        "brisket",
    },
    "cheese": {
        "cheese",
        "cheddar",
        "colby jack",
        "provolone",
        "queso fresco",
        "pepper jack",
        "parmigiano",
        "mascarpone",
        "fontina",
        "pecorino",
        "feta",
        "parmesan",
        "gruyère",
        "gruyere",
        "mozzarella",
    },
    "dairy": {
        "egg",
        "milk",
        "buttermilk",
        "cream",
        "ricotta",
        "yogurt",
        "yoghurt",
        "coconut noosa",
        "butter",
        "half-and-half",
        "creme",
        "creme fraiche",
        "crème fraîche",
    },
    "baking": {
        "sugar",
        "granulated",
        "flour",
        "almond flour",
        "chocolate",
        "choclate",
        "poppy seeds",
        "molasses",
        "lard",
        "ghee",
        "gee",
        "cocoa",
        "food coloring",
        "cocoa powder",
        "coca powder",
        "baking soda",
        "baking powder",
        "vegetable shortening",
        "malted milk powder",
        "corn syrup",
        "coconut sugar",
    },
    "frozen": {
        "bag potstickers",
        "potstickers",
        "phyllo dough",
        "phyllo pastry",
        "puff pastry",
        "pie dough",
        "pie crust",
        "pie shell",
        "frozen gyoza",
    },
    "bread": {
        "bread",
        "panettone",
        "pita",
        "naan",
        "bun",
        "loaf",
        "burger bun",
        "buns",
        "challah",
        "brioche",
        "rolls",
        "baguette",
        "tortillas",
        "barley rusks",
        "sub roll",
        "english muffins",
        "bagel",
        "hot dog buns",
        "pizza dough",
    },
    "canned & packaged": {
        "anchovy",
        "fruit preserves",
        "chickpeas",
        "beans",
        "sardines",
        "pumpkin puree",
        "papaya paste",
        "canned tomatoes",
        "crushed tomatoes",
        "can whole tomatoes",
        "can whole peeled tomatoes",
        "cans of diced tomatoes",
        "can San Marzano tomatoes",
        "pumpkin purée",
        "marinara sauce",
        "arrabbiata sauce",
        "oyster sauce",
        "tomato paste",
        "tuna",
        "anchovies",
        "adobo",
        "broth",
        "chicken broth",
        "stock",
        "coconut milk",
        "cream of coconut",
        "chicken stock",
        "beef stock",
        "fish stock",
        "vegetable stock",
    },
    # kind of the misc category
    "condiments": {
        "tamari",
        "pesto",
        "horseradish",
        "peanut butter",
        "almond butter",
        "doubanjiang",
        "fish sauce",
        "gochujang",
        "pickled jalapeno",
        "salsa",
        "chimichurri",
        "pickle",
        "pizza sauce",
        "aioli",
        "mayo",
        "mayonnaise",
        "mango chutney",
        "marshmallows",
        "ketchup",
        "ketcup",
        "oil",
        "olive oil",
        "coconut oil",
        "peanut oil",
        "honey",
        "dulce de leche",
        "hot sauce",
        "capers",
        "vinegar",
        "soy sauce",
        "tamarind",
        "tahini",
        "tzatziki",
        "sprinkles",
        "sesame paste",
        "mustard",
        "seaweed",
        "jam",
        "maple syrup",
        "marmite",
        "sriracha",
        "ya cai",
        "yacai",
        "curry paste",
        "worcestershire sauce",
        "bean paste",
        "almond paste",
        "agave nectar",
        "olives",
    },
    "dry goods": {
        "gingersnaps",
        "pepitas",
        "currants",
        "powdered msg",
        "graham cracker",
        "tortilla chips",
        "breadcrumbs",
        "quinoa",
        "raisins",
        "matcha",
        "masa harina",
        "polenta",
        "cornmeal",
        "panko",
        "orzo",
        "oatmeal",
        "oats",
        "lentils",
        "ladyfingers",
        "nilla wafers",
        "granola",
        "fritos",
        "bulgar",
        "bulgur",
        "millet",
        "flax seed",
        "coffee",
        "espresso",
        "chia",
        "couscous",
        "gelatin",
        "amaranth",
        "katsuobushi",
        "ao nori",
        "barley",
        "farro",
        "ramen",
        "fettuccine",
        "macaroni",
        "fusilli",
        "rotini",
        "penne",
        "rigatoni",
        "yeast",
        "rice",
        "pasta",
        "lasagna noodles",
        "tapioca",
        "cornstarch",
        "noodles",
        "spaghetti",
        "pastina",
        "split peas",
    },
    "spices": {
        "asafetida",
        "cinnamon",
        "curry powder",
        "sumac",
        "peppercorns",
        "za'atar",
        "paprika",
        "salt",
        "furikake",
        "caraway seeds",
        "bati masala",
        "masala",
        "five-spice powder",
        "mustard powder",
        "garlic powder",
        "allspice",
        "onion powder",
        "dried oregano",
        "ground ginger",
        "pepper",
        "bay leaves",
        "miso",
        "cloves",
        "fennel seed",
        "ground clove",
        "red chile flakes",
        "chile flakes",
        "chili flakes",
        "chili powder",
        "chilli powder",
        "chile powder",
        "cumin",
        "coriander",
        "corainder",
        "vanilla",
        "whole clove",
        "nutmeg",
        "turmeric",
        "sesame seed",
        "mustard seed",
        "dried chilies",
        "dried chiles",
        "fenugreek",
        "ground chiles",
        "baharat",
        "seasoning",
        "star anise",
        "marjoram",
        "bay leaf",
        "vanilla extract",
        "lemon extract",
        "almond extract",
        "cardamom",
        "cayenne",
        "cream of tartar",
        "garam masala",
        "herbes de provence",
        "saffron",
        "niger seed",
        "nigella seed",
        "peperoncino flakes",
    },
    "nuts": {
        "pecans",
        "pecan",
        "pine nuts",
        "peanuts",
        "walnut",
        "almonds",
        "nuts",
        "cashews",
        "pistachios",
    },
    "other": {
        "cedar shakes",
        "water",
        "ice",
        "gloves",
        "skewers",
    },
}


@cache
def _create_trie() -> dict[str, Any]:
    trie: dict[str, Any] = {}
    for category, ingredients in _DEPARTMENT_MAPPING.items():
        for ingredient in ingredients:
            tree = trie
            words = [singularize(x) for x in ingredient.replace("-", " ").split()]
            for idx, word in enumerate(words):
                if word in tree:
                    tree = tree[word]
                else:
                    tree[word] = {}
                    tree = tree[word]
                is_last = idx == len(words) - 1
                if is_last and "$" not in tree:
                    tree["$"] = (category, len(ingredient))

    return trie


def _search(item: str) -> dict[str, set[int]]:
    trie = _create_trie()
    items = [singularize(x) for x in item.split()]
    counts = defaultdict(set)
    for start in range(len(items)):
        tree = trie
        for word in items[start:]:
            if word not in tree:
                break
            tree = tree[word]
            if "$" in tree:
                cat, cnt = tree["$"]
                counts[cat].add(cnt)
    return counts


def category(ingredient: str) -> str:
    res = _search(
        ingredient.lower()
        .replace("-", " ")
        .replace(",", "")
        .replace(")", "")
        .replace("(", "")
        .replace("’", "'")  # noqa: RUF001
    )
    if not res:
        return "unknown"

    return sorted(res.items(), key=lambda x: -max(x[1]))[0][0]
