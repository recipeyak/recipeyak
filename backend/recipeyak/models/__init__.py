from __future__ import annotations

from django.db.models import QuerySet
from django.shortcuts import get_object_or_404

from recipeyak.models.calendar import Calendar  # noqa: F401
from recipeyak.models.ingredient import Ingredient
from recipeyak.models.ingredient_historical import IngredientHistorical  # noqa: F401
from recipeyak.models.invite import Invite  # noqa: F401
from recipeyak.models.membership import Membership, get_random_ical_id  # noqa: F401
from recipeyak.models.note import Note
from recipeyak.models.note_historical import NoteHistorical as NoteHistorical
from recipeyak.models.reaction import Reaction
from recipeyak.models.recipe import Recipe
from recipeyak.models.recipe_change import ChangeType, RecipeChange  # noqa: F401
from recipeyak.models.recipe_cook_checklist_check import (
    RecipeCookChecklistCheck,
)
from recipeyak.models.recipe_favorite import RecipeFavorite  # noqa: F401
from recipeyak.models.recipe_historical import RecipeHistorical  # noqa: F401
from recipeyak.models.recipe_index_queue import RecipeIndexQueue  # noqa: F401
from recipeyak.models.recipe_view import RecipeView  # noqa: F401
from recipeyak.models.schedule_event import ScheduleEvent  # noqa: F401
from recipeyak.models.scheduled_recipe import ScheduledRecipe  # noqa: F401
from recipeyak.models.scrape import Scrape  # noqa: F401
from recipeyak.models.section import Section  # noqa: F401
from recipeyak.models.section_historical import SectionHistorical  # noqa: F401
from recipeyak.models.shopping_list import ShoppingList  # noqa: F401
from recipeyak.models.step import Step
from recipeyak.models.step_historical import StepHistorical  # noqa: F401
from recipeyak.models.team import Team
from recipeyak.models.timeline_event import TimelineEvent  # noqa: F401
from recipeyak.models.upload import Upload
from recipeyak.models.user import User


def get_team(user: User) -> Team:
    # HACK: ideally we'd have this accessible on the request (request.team) but
    # with DRF it's close to impossible to access authed user
    # see: https://stackoverflow.com/questions/26240832/django-and-middleware-which-uses-request-user-is-always-anonymous

    # Also, we should send the team with the request. `user.schedule_team`
    # should only be used to populate the default team value on login.
    # If we send it with the request, then a user can have multiple tabs work
    # properly.
    team_id = user.schedule_team_id
    assert team_id is not None, "should always have a team selected"
    return get_team_by_id(user_id=user.id, team_id=team_id)


def get_pinned_calendar(user: User, team_id: int) -> Calendar:
    calendar_id = user.pinned_calendar_id
    assert calendar_id is not None, "should always have a pinned calendar selected"
    return Calendar.objects.filter(team=team_id).get(id=calendar_id)


def get_team_by_id(*, user_id: int, team_id: int) -> Team:
    return get_object_or_404(
        Team.objects.filter(
            membership__user_id=user_id,
            membership__is_active=True,
        ),
        id=team_id,
    )


def filter_recipes(*, team: Team) -> QuerySet[Recipe]:
    return Recipe.objects.filter(team=team).prefetch_related(
        "step_set",
        "ingredient_set",
        "scheduledrecipe_set",
        "notes",
        "notes__created_by",
        "notes__created_by__profile_upload",
        "notes__uploads",
        "notes__reactions",
        "notes__reactions__created_by",
        "timelineevent_set",
        "timelineevent_set__created_by",
        "timelineevent_set__created_by__profile_upload",
        "section_set",
        "primary_image",
        "primary_image__created_by",
        "primary_image__recipe",
    )


def filter_recipe_or_404(*, recipe_id: int, team: Team) -> Recipe:
    return get_object_or_404(filter_recipes(team=team), pk=recipe_id)


def filter_cook_checklist(*, team: Team) -> QuerySet[RecipeCookChecklistCheck]:
    return RecipeCookChecklistCheck.objects.filter(recipe__in=filter_recipes(team=team))


def filter_ingredients(*, team: Team) -> QuerySet[Ingredient]:
    return Ingredient.objects.filter(recipe__in=filter_recipes(team=team))


def filter_steps(*, team: Team) -> QuerySet[Step]:
    return Step.objects.filter(recipe__in=filter_recipes(team=team))


def filter_notes(*, team: Team) -> QuerySet[Note]:
    return Note.objects.filter(recipe__in=filter_recipes(team=team))


def filter_uploads(*, team: Team, user_id: str) -> QuerySet[Upload]:
    return Upload.objects.filter(
        recipe__in=filter_recipes(team=team), created_by=user_id
    )


def user_reactions(*, user: User) -> QuerySet[Reaction]:
    return Reaction.objects.filter(created_by=user)
