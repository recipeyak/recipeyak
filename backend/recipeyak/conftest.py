from __future__ import annotations

from collections.abc import Iterator
from datetime import date
from unittest.mock import patch

import pytest
from django.test.client import Client

from recipeyak.models import (
    Ingredient,
    Note,
    Recipe,
    ScheduledRecipe,
    Section,
    Step,
    Team,
    User,
)
from recipeyak.models.calendar import Calendar


@pytest.fixture
def user() -> User:
    """
    Connected to `team`. Has 5 recipes. Member of `team`.
    """
    email = "john@doe.org"
    return User.objects.create_user(email=email)


@pytest.fixture
def user2() -> User:
    email = "james@smith.org"
    user2 = User.objects.create_user(email=email)
    team = Team.objects.create(name="another team")
    team.force_join_admin(user=user2)
    user2.schedule_team = team
    user2.save()
    return user2


@pytest.fixture
def user3() -> User:
    email = "john.doe@example.org"
    user3 = User.objects.create_user(email=email)
    team = Team.objects.create(name="another team")
    team.force_join_admin(user=user3)
    user3.schedule_team = team
    user3.save()
    return user3


@pytest.fixture
def client() -> Client:
    return Client()


@pytest.fixture
def client_b() -> Client:
    return Client()


@pytest.fixture
def recipes(user: User, team: Team) -> list[Recipe]:
    """
    list of empty recipes with different modified times owned by `user`
    note: each item will have a different creation time inherently
    """

    name = "Recipe name"
    author = "Recipe author"

    return Recipe.objects.bulk_create(
        [Recipe(name=name, author=author, team=team) for n in range(15)]
    )


@pytest.fixture
def recipe(user: User, team: Team) -> Recipe:
    """
    Recipe with metadata, Ingredient, Step owned by `user`
    """

    name = "Recipe name"
    author = "Recipe author"
    source = "www.exmple.com"
    time = "1 hour"

    recipe = Recipe.objects.create(
        name=name, author=author, source=source, time=time, team=team
    )

    Ingredient.objects.create(
        quantity="1 lbs",
        name="egg",
        description="scrambled",
        position=10.0,
        recipe=recipe,
    )

    Ingredient.objects.create(
        quantity="2 tbs", name="soy sauce", position=11.0, recipe=recipe
    )

    Step.objects.create(
        text="Place egg in boiling water and cook for ten minutes",
        position=10.0,
        recipe=recipe,
    )

    Note.objects.create(recipe=recipe, text="Use a small pot.", created_by=user)
    Note.objects.create(
        recipe=recipe,
        text="I'd say we should use an even smaller pot next time.",
        created_by=user,
        last_modified_by=user,
    )

    Section.objects.create(recipe=recipe, title="a section title", position=14.0)
    Section.objects.create(recipe=recipe, title="a diff section", position=5.5)

    return recipe


def recipe_pie_factory(user: User, team: Team) -> Recipe:
    """
    Recipe with metadata, Ingredient, Step owned by `user`.
    Contains the word "pie" in name and source.
    """
    name = "Brandied Pumpkin Pie"
    author = "Melissa Clark"
    source = "https://cooking.nytimes.com/recipes/1015413-brandied-pumpkin-pie"
    time = "4 hours"

    recipe = Recipe.objects.create(
        name=name, author=author, source=source, time=time, team=team
    )

    Ingredient.objects.create(
        quantity="150 grams", name="all-purpose flour", position=10.0, recipe=recipe
    )

    Ingredient.objects.create(
        quantity="0.25 teaspoon", name="fine sea salt", position=11.0, recipe=recipe
    )

    Step.objects.create(
        text="Make the crust: In a food processor, pulse together the flour and salt.",
        position=10.0,
        recipe=recipe,
    )

    return recipe


@pytest.fixture
def recipe_pie(user: User, team: Team) -> Recipe:
    return recipe_pie_factory(user, team)


@pytest.fixture
def recipe2(user: User, team: Team) -> Recipe:
    """
    Pie recipe owned by `user`
    """
    return recipe_pie_factory(user, team)


@pytest.fixture
def team(user: User) -> Team:
    team = Team.objects.create(name="Recipe Yak Team")
    calendar = Calendar.objects.create(team=team, name="Default")
    team.force_join_admin(user=user)
    user.schedule_team = team
    user.pinned_calendar = calendar
    user.save()
    return team


@pytest.fixture
def empty_team() -> Team:
    return Team.objects.create(name="The Hateful Eight")


@pytest.fixture
def team_with_recipes(team: Team, recipe: Recipe, recipe_pie: Recipe) -> Team:
    team.recipe_set.add(recipe)
    team.recipe_set.add(recipe_pie)
    return team


@pytest.fixture
def team_with_recipes_no_members(recipe: Recipe, recipe_pie: Recipe) -> Team:
    t = Team.objects.create(name="A Team with No Name")
    t.recipe_set.add(recipe)
    t.recipe_set.add(recipe_pie)
    return t


@pytest.fixture
def scheduled_recipe(recipe: Recipe, team: Team) -> ScheduledRecipe:
    return ScheduledRecipe.objects.create(recipe=recipe, team=team, on=date(1976, 7, 6))


@pytest.fixture(scope="session", autouse=True)
def patch_publish_calendar_event() -> Iterator[None]:
    with (
        patch("recipeyak.realtime.publish_calendar_event", return_value=None),
        patch("recipeyak.realtime.publish_calendar_event_deleted", return_value=None),
    ):
        yield


@pytest.fixture(scope="session", autouse=True)
def patch_publish_cook_checklist() -> Iterator[None]:
    with patch("recipeyak.realtime.publish_cook_checklist", return_value=None):
        yield


@pytest.fixture(scope="session", autouse=True)
def patch_publish_recipe_modified() -> Iterator[None]:
    with patch("recipeyak.realtime.publish_recipe", return_value=None):
        yield
