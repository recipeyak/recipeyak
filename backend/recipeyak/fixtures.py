"""
Create test data
"""

from recipeyak.models.calendar import Calendar
from recipeyak.models.ingredient import Ingredient
from recipeyak.models.note import Note
from recipeyak.models.recipe import Recipe
from recipeyak.models.section import Section
from recipeyak.models.step import Step
from recipeyak.models.team import Team
from recipeyak.models.user import User
from recipeyak.versioning import save_recipe_version


def create_user(*, email: str = "j.person@example.com") -> User:
    return User.objects.create_user(email=email)


def create_team(*, user: User) -> Team:
    team = Team.objects.create(name="Recipe Yak Team")
    calendar = Calendar.objects.create(team=team, name="Default")
    team.force_join_admin(user=user)
    user.schedule_team = team
    user.pinned_calendar = calendar
    user.save()
    return team


def create_recipe(*, team: Team, user: User, source: str | None = None) -> Recipe:
    recipe = Recipe.objects.create(
        name="Recipe name",
        author="Recipe author",
        source=source or "www.exmple.com",
        time="1 hour",
        team=team,
    )

    Section.objects.create(recipe=recipe, title="a section title", position="a")
    create_ingredient(
        recipe=recipe,
        quantity="1 lbs",
        name="egg",
        description="scrambled",
        position="b",
    )

    Section.objects.create(recipe=recipe, title="a diff section", position="b")

    create_ingredient(
        recipe=recipe,
        quantity="2 tbs",
        name="soy sauce",
        position="c",
    )

    Step.objects.create(
        text="Place egg in boiling water and cook for ten minutes",
        position="a",
        recipe=recipe,
    )

    Step.objects.create(
        text="serve",
        position="b",
        recipe=recipe,
    )

    Note.objects.create(recipe=recipe, text="Use a small pot.", created_by=user)
    Note.objects.create(
        recipe=recipe,
        text="I'd say we should use an even smaller pot next time.",
        created_by=user,
        last_modified_by=user,
    )

    # mimic what we do in the API
    save_recipe_version(recipe_id=recipe.id, actor=user)

    return recipe


def create_ingredient(
    *, recipe: Recipe, quantity: str, name: str, description: str = "", position: str
) -> Ingredient:
    return Ingredient.objects.create(
        quantity=quantity,
        name=name,
        description=description,
        recipe=recipe,
        position=position,
    )
