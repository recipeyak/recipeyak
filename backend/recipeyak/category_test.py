import csv
from collections import defaultdict
from pathlib import Path
from textwrap import dedent

from syrupy.assertion import SnapshotAssertion

from recipeyak.category import _DEPARTMENT_MAPPING, category


def test_categorize_ingredients() -> None:
    """
    Ensure we have a category set for each ingredient
    """
    out = defaultdict(set)
    with (Path() / "ingredients.csv").open() as f:
        reader = csv.DictReader(f)
        for row in reader:
            item = row["name"]
            cat = category(item.lower())
            out[cat].add(item)

    expected_unknown = {
        "for a 9-inch double crust pie (see other recipe)",
        "thinly sliced",
    }

    out["unknown"] -= expected_unknown

    assert not out["unknown"], f'remaining {len(out["unknown"])}'


def test_categorize_ingredient_test_cases(snapshot: SnapshotAssertion) -> None:
    cases = dedent(
        """
    (13-ounce) can unsweetened coconut milk
    (14.5-ounce) can whole or crushed tomatoes
    (15-ounce) cans chickpeas
    (6-inch) corn tortillas
    1 lime
    1-inch piece ginger
    1/4 ounce active dry yeast
    14 oz cans of diced tomatoes
    14.5-ounce can whole peeled tomatoes
    15-ounce can chickpeas
    15-ounce container ricotta cheese
    28 ounce can San Marzano tomatoes
    28-ounce can whole tomatoes
    3-to-3.5-pound butternut squash
    4-ounce milk chocolate bar
    Asian fish sauce
    Chinese rice wine
    Grated Parmigiano-Reggiano
    all-butter puff pastry
    almond butter
    almond paste
    pickled jalapeno
    pepper jack cheese
    apple cider vinegar
    coarsely grated mozzarella
    small basil leaves (preferably thai or cinnamon basil)
    coconut noosa
    almond flour
    baby bok choy
    beef broth
    beef or chicken broth
    beef stock
    best-quality apricot jam or orange marmalade
    bulk hot or sweet breakfast sausage or Italian sausage
    butternut squash
    butternut squash (1 small squash)
    chicken stock
    chinese egg noodles
    coconut extract
    coconut oil
    coconut rum or dark rum
    coconut sugar or substitute brown sugar
    corn torillas
    corn tortilla chips
    corn tortillas
    cornmeal
    cornstarch
    cream of coconut
    cream of tartar
    creamy unsweetened peanut butter
    crystallized ginger
    cubanelle peppers
    dark corn syrup
    diced canned tomatoes
    dill pickle chips
    doubanjiang (Sichuanese fermented chile bean paste; see Note)
    dried green split peas
    dry mustard powder
    eggplant
    finely chopped cornichons or small kosher dill pickles
    finely grated ginger
    fish sauce
    barley rusks
    slices bread
    boti masala
    cloves garlic
    garlic cloves
    dried oregano
    onion powder
    graham cracker crumbs
    fennel seed
    ground ginger
    hamburger buns
    haricot vert fins or green beans
    harissa or tomato paste
    hot italian sausage
    large poblano pepper
    lemon extract
    red bird eye chillies, roughly chopped
    light corn syrup
    store bought or homemade zaâ€™atar
    loaf sweet egg bread like challah or brioche
    oyster sauce
    fresno peppers
    cherry or grape tomatoes
    matcha powder
    sesame aioli
    raw papaya paste
    (3/4 inch thick) slices of a 10 inch round peasant loaf
    chipotle aioli (store bought)
    low sodium chicken stock
    low-sodium chicken broth
    malted milk powder
    medium-coarse yellow cornmeal
    tomato paste
    """  # noqa: RUF001
    ).splitlines()

    assert sorted((w, category(w)) for w in cases) == snapshot()


def test_no_overlap_between_categories() -> None:
    mapping = {key: set(value) for key, value in _DEPARTMENT_MAPPING.items()}

    for key, value in mapping.items():
        for other_key, other_value in mapping.items():
            if key == other_key:
                continue
            overlap = value & other_value
            assert not overlap, f"{overlap} in {key} and {other_key}"
