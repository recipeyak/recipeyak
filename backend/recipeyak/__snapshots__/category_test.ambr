# name: test_categorize_ingredient_test_cases
  list([
    tuple(
      '',
      'unknown',
    ),
    tuple(
      '(13-ounce) can unsweetened coconut milk',
      'canned & packaged',
    ),
    tuple(
      '(14.5-ounce) can whole or crushed tomatoes',
      'canned & packaged',
    ),
    tuple(
      '(15-ounce) cans chickpeas',
      'canned & packaged',
    ),
    tuple(
      '(3/4 inch thick) slices of a 10 inch round peasant loaf',
      'bread',
    ),
    tuple(
      '(6-inch) corn tortillas',
      'bread',
    ),
    tuple(
      '1 lime',
      'produce',
    ),
    tuple(
      '1-inch piece ginger',
      'produce',
    ),
    tuple(
      '1/4 ounce active dry yeast',
      'dry goods',
    ),
    tuple(
      '14 oz cans of diced tomatoes',
      'canned & packaged',
    ),
    tuple(
      '14.5-ounce can whole peeled tomatoes',
      'canned & packaged',
    ),
    tuple(
      '15-ounce can chickpeas',
      'canned & packaged',
    ),
    tuple(
      '15-ounce container ricotta cheese',
      'dairy',
    ),
    tuple(
      '28 ounce can San Marzano tomatoes',
      'produce',
    ),
    tuple(
      '28-ounce can whole tomatoes',
      'canned & packaged',
    ),
    tuple(
      '3-to-3.5-pound butternut squash',
      'produce',
    ),
    tuple(
      '4-ounce milk chocolate bar',
      'baking',
    ),
    tuple(
      'Asian fish sauce',
      'condiments',
    ),
    tuple(
      'Chinese rice wine',
      'dry goods',
    ),
    tuple(
      'Grated Parmigiano-Reggiano',
      'cheese',
    ),
    tuple(
      'all-butter puff pastry',
      'frozen',
    ),
    tuple(
      'almond butter',
      'condiments',
    ),
    tuple(
      'almond flour',
      'baking',
    ),
    tuple(
      'almond paste',
      'condiments',
    ),
    tuple(
      'apple cider vinegar',
      'condiments',
    ),
    tuple(
      'baby bok choy',
      'produce',
    ),
    tuple(
      'barley rusks',
      'bread',
    ),
    tuple(
      'beef broth',
      'canned & packaged',
    ),
    tuple(
      'beef or chicken broth',
      'canned & packaged',
    ),
    tuple(
      'beef stock',
      'canned & packaged',
    ),
    tuple(
      'best-quality apricot jam or orange marmalade',
      'produce',
    ),
    tuple(
      'boti masala',
      'spices',
    ),
    tuple(
      'bulk hot or sweet breakfast sausage or Italian sausage',
      'meat',
    ),
    tuple(
      'butternut squash',
      'produce',
    ),
    tuple(
      'butternut squash (1 small squash)',
      'produce',
    ),
    tuple(
      'cherry or grape tomatoes',
      'produce',
    ),
    tuple(
      'chicken stock',
      'canned & packaged',
    ),
    tuple(
      'chinese egg noodles',
      'dry goods',
    ),
    tuple(
      'chipotle aioli (store bought)',
      'condiments',
    ),
    tuple(
      'cloves garlic',
      'produce',
    ),
    tuple(
      'coarsely grated mozzarella',
      'cheese',
    ),
    tuple(
      'coconut extract',
      'produce',
    ),
    tuple(
      'coconut noosa',
      'dairy',
    ),
    tuple(
      'coconut oil',
      'condiments',
    ),
    tuple(
      'coconut rum or dark rum',
      'produce',
    ),
    tuple(
      'coconut sugar or substitute brown sugar',
      'baking',
    ),
    tuple(
      'corn torillas',
      'produce',
    ),
    tuple(
      'corn tortilla chips',
      'dry goods',
    ),
    tuple(
      'corn tortillas',
      'bread',
    ),
    tuple(
      'cornmeal',
      'dry goods',
    ),
    tuple(
      'cornstarch',
      'dry goods',
    ),
    tuple(
      'cream of coconut',
      'canned & packaged',
    ),
    tuple(
      'cream of tartar',
      'spices',
    ),
    tuple(
      'creamy unsweetened peanut butter',
      'condiments',
    ),
    tuple(
      'crystallized ginger',
      'produce',
    ),
    tuple(
      'cubanelle peppers',
      'produce',
    ),
    tuple(
      'dark corn syrup',
      'baking',
    ),
    tuple(
      'diced canned tomatoes',
      'canned & packaged',
    ),
    tuple(
      'dill pickle chips',
      'condiments',
    ),
    tuple(
      'doubanjiang (Sichuanese fermented chile bean paste; see Note)',
      'condiments',
    ),
    tuple(
      'dried green split peas',
      'dry goods',
    ),
    tuple(
      'dried oregano',
      'spices',
    ),
    tuple(
      'dry mustard powder',
      'spices',
    ),
    tuple(
      'eggplant',
      'produce',
    ),
    tuple(
      'fennel seed',
      'spices',
    ),
    tuple(
      'finely chopped cornichons or small kosher dill pickles',
      'condiments',
    ),
    tuple(
      'finely grated ginger',
      'produce',
    ),
    tuple(
      'fish sauce',
      'condiments',
    ),
    tuple(
      'fresno peppers',
      'produce',
    ),
    tuple(
      'garlic cloves',
      'produce',
    ),
    tuple(
      'graham cracker crumbs',
      'dry goods',
    ),
    tuple(
      'ground ginger',
      'spices',
    ),
    tuple(
      'hamburger buns',
      'bread',
    ),
    tuple(
      'haricot vert fins or green beans',
      'produce',
    ),
    tuple(
      'harissa or tomato paste',
      'canned & packaged',
    ),
    tuple(
      'hot italian sausage',
      'meat',
    ),
    tuple(
      'large poblano pepper',
      'produce',
    ),
    tuple(
      'lemon extract',
      'spices',
    ),
    tuple(
      'light corn syrup',
      'baking',
    ),
    tuple(
      'loaf sweet egg bread like challah or brioche',
      'bread',
    ),
    tuple(
      'low sodium chicken stock',
      'canned & packaged',
    ),
    tuple(
      'low-sodium chicken broth',
      'canned & packaged',
    ),
    tuple(
      'malted milk powder',
      'baking',
    ),
    tuple(
      'matcha powder',
      'dry goods',
    ),
    tuple(
      'medium-coarse yellow cornmeal',
      'dry goods',
    ),
    tuple(
      'onion powder',
      'spices',
    ),
    tuple(
      'oyster sauce',
      'canned & packaged',
    ),
    tuple(
      'pepper jack cheese',
      'cheese',
    ),
    tuple(
      'pickled jalapeno',
      'condiments',
    ),
    tuple(
      'raw papaya paste',
      'canned & packaged',
    ),
    tuple(
      'red bird eye chillies, roughly chopped',
      'produce',
    ),
    tuple(
      'sesame aioli',
      'condiments',
    ),
    tuple(
      'slices bread',
      'bread',
    ),
    tuple(
      'small basil leaves (preferably thai or cinnamon basil)',
      'produce',
    ),
    tuple(
      'store bought or homemade zaâ€™atar',
      'spices',
    ),
    tuple(
      'tomato paste',
      'canned & packaged',
    ),
  ])
# ---
